python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,108 @@
+class Trainer(ConvertibleToTrainable, abc.ABC):
+    """
+    Defines interface for distributed training on Ray.
+
+    Args:
+        scaling_config (Optional[ScalingConfig]): Configuration for how to
+            scale training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or a Callbable that returns a Dataset,
+            to use for training. If a ``preprocessor`` is also provided,
+            it will be fit on this dataset and this dataset will be
+            transformed.
+        additional_datasets (Optional[GenDataset]): Any additional
+            Datasets (such as validation or test datasets) to use for
+            training. If a ``preprocessor`` is provided, the datasets
+            specified here will only be transformed, and not fit on.
+        preprocessor (Optional[Preprocessor]): A preprocessor to preprocess
+            the provided datasets.
+        resume_from_checkpoint (Optional[Checkpoint]): A checkpoint to
+            resume training from.
+    """
+
+    def __init__(
+        self,
+        scaling_config: Optional[ScalingConfig] = None,
+        run_config: Optional[RunConfig] = None,
+        train_dataset: Optional[GenDataset] = None,
+        additional_datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional[Preprocessor] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        self.scaling_config = scaling_config
+        self.run_config = run_config if run_config else {}
+        self.train_dataset = train_dataset
+        self.additional_datasets = additional_datasets
+        self.preprocessor = preprocessor
+        self.checkpoint_to_resume_from = resume_from_checkpoint
+
+    def fit(self) -> Result:
+        """Runs training.
+
+        Returns:
+            A Result object containing the training result.
+        """
+        trainable = self.as_trainable()
+
+        # Copied from initial prototyping.
+        # TODO: Replace with Tuner.
+        analysis = tune.run(run_or_experiment=trainable, **self.run_config)
+
+        assert len(analysis.trials) == 1
+
+        trial = analysis.trials[0]
+
+        result = Result(
+            metrics=trial.last_result,
+            checkpoint=Checkpoint.from_directory(trial.checkpoint.value)
+            if trial.checkpoint.value
+            else None,
+        )
+
+        return result
+
+    def _override_attributes_with_config(self, config: Dict):
+        """Overrides attributes of the Trainer with values in ``config``.
+
+        This is needed to allow attributes of Trainer to be tuned as
+        hyperparameters.
+
+        This method does a deep update of dictionaries.
+
+        Args:
+            config (Dict): A dictionary to update attributes with.
```

Comment:
```suggestion
            config: A dictionary to update attributes with.
```

Please make this change here and elsewhere.

Reply:
Fixed.